File: lib/js/MetaData.js
#
# JavaScript Tools - Meta Data Loader
# Copyright 2010 Sebastian Werner
#

#
# Public API
#

def collect(node, ownName=None):
    """ Computes and returns the dependencies of the given node """

    return __inspect(node, MetaData())



#
# Implementation
#

class MetaData:
    """ Data structure to hold all dependency information """
    def __init__(self):
        self.provides = set()
        self.requires = set()
        self.optionals = set()
        self.breaks = set()
        
        
def __inspect(node, meta):
    """ The internal inspection routine """
    
    # Parse comments
    try:
        comments = node.comments
    except AttributeError:
        comments = None
    
    if comments:
        for comment in comments:
            commentTags = comment.getTags()
            if commentTags:
                if "provide" in commentTags:
                    meta.provides.update(set(commentTags["provide"]))
                if "require" in commentTags:
                    meta.requires.update(set(commentTags["require"]))
                if "optional" in commentTags:
                    meta.optionals.update(set(commentTags["optional"]))
                if "break" in commentTags:
                    meta.breaks.update(set(commentTags["break"]))

    # Process children
    for child in node:
        __inspect(child, meta)
        
    return data

File: jasy/js/MetaData.py
#
# Jasy - Web Tooling Framework
# Copyright 2010-2012 Zynga Inc.
#

class MetaData:
    """ 
    Data structure to hold all meta information. 
    
    A instance of this class is typically created by processing all 
    meta data relevant tags of all doc comments in the given node structure.

    Hint: Must be a clean data class without links to other 
    systems for optiomal cachability using Pickle
    """
    
    __slots__ = ["name", "requires", "optionals", "breaks", "assets"]
    
    def __init__(self, tree):
        self.name = None
        
        self.requires = set()
        self.optionals = set()
        self.breaks = set()
        self.assets = set()
        
        self.__inspect(tree)
        
        
    def __inspect(self, node):
        """ The internal inspection routine """
    
        # Parse comments
        comments = getattr(node, "comments", None)
        if comments:
            for comment in comments:
                commentTags = comment.getTags()
                if commentTags:

                    if "name" in commentTags:
                        self.name = list(commentTags["name"])[0]
                    if "require" in commentTags:
                        self.requires.update(commentTags["require"])
                    if "load" in commentTags:
                        # load is a special combination shorthand for requires + breaks
                        # This means load it but don't require it being loaded first
                        self.requires.update(commentTags["load"])
                        self.breaks.update(commentTags["load"])
                    if "optional" in commentTags:
                        self.optionals.update(commentTags["optional"])
                    if "break" in commentTags:
                        self.breaks.update(commentTags["break"])
                    if "asset" in commentTags:
                        self.assets.update(commentTags["asset"])

        # Process children
        for child in node:
            if child is not None:
                self.__inspect(child)
